src/jit/flowgraph.cpp

```diff
@@ -22337,6 +22337,39 @@ GenTreePtr Compiler::fgInlinePrependStatements(InlineInfo* inlineInfo)
                 }
                 else
                 {
+                    // We're going to assign the argument value to the
+                    // temp we use for it in the inline body.
+                    //
+                    // If we know the argument's value can't be
+                    // changed within the method body, try and improve
+                    // the type of the temp.
+                    if (!inlArgInfo[argNum].argHasLdargaOp && !inlArgInfo[argNum].argHasStargOp)
+                    {
+                        GenTree*             argNode        = inlArgInfo[argNum].argNode;
+                        bool                 isExact        = false;
+                        bool                 isNonNull      = false;
+                        CORINFO_CLASS_HANDLE refClassHandle = gtGetClassHandle(argNode, &isExact, &isNonNull);
+
+                        if (refClassHandle != nullptr)
+                        {
+                            const unsigned tmpNum = inlArgInfo[argNum].argTmpNum;
+
+                            // If we already had an exact type for
+                            // this temp, this new information had
+                            // better agree with what we knew before.
+                            if (lvaTable[tmpNum].lvClassIsExact)
+                            {
+                                assert(isExact);
+                                assert(refClassHandle == lvaTable[tmpNum].lvClassHnd);
```

Comment:
This is a bit of future proofing.

Currently the argument temps -- created over in `impInlineFetchArg` -- don't get any class/exact info. If/when they do, we will type them based on the method signature.

Any correct program will pass an argument value with the signature type or some subtype. So the argument type should be as good or better a bound on the type than the signature type.

In the update case, it might be worth checking for proper subtype relationship here, though if we somehow "downgrade" the type to a superclass, we still have a safeguard in the VM side of devirtualization, which also does this check.